Vendoring-compatible imports

jku opened the pull request:
The goal here is to make TUF compatible with the vendoring tool so that tuf sources can be vendored into pip. As a reminder vendoring will modify all import-lines in a vendored project like this patch:

```
- import package 
- from package import module 
- from package.module import func
+ from _vendoring import package 
+ from _vendoring.package import module 
+ from _vendoring.package.module import func

```

This does not work in the case of `import package.module` that is so common in tuf.

Notes
* tuf modules are now imported with this style `from tuf import download`. This works with just a few localised issues: I'm generally happy with how code calling internal modules looks
* securesystemslib modules can't be handled this way:
  * module names clash with tuf modules names
  * module names are quite generic and likely to lead to conflicts later on
* as result securesystemslib imports are now like `from securesystemslib import formats as sslib_formats` in most cases. I don't particularly like it but it is readable and I can't think of a better solution
* Very minor changes to how other dependencies are used
* linter now understands our imports so some fixes and workarounds have been added. 
* The import lists that we're changed have been sorted so tuf imports are last
* The patch is surprisingly large, ~900 lines. Roughly 350 are needed for the client vendoring but I modified the whole code base to at least be consistent. The tests have not been touched.

Commit:
imports: Make 'download' imports vendoring-compatible

Use "from tuf import <module>" instead of "import tuf.<module>": this
makes it possible for vendoring tool to vendor tuf. Fix all references
to <module> in the code.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Make 'exceptions' imports vendoring-compatible

Use "from tuf import <module>" instead of "import tuf.<module>": this
makes it possible for vendoring tool to vendor tuf. Fix all references
to <module> in the code.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Make 'formats' imports vendoring-compatible

Use "from tuf import <module>" instead of "import tuf.<module>": this
makes it possible for vendoring tool to vendor tuf. Fix all references
to <module> in the code.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Make 'keydb' imports vendoring-compatible

Use "from tuf import <module>" instead of "import tuf.<module>": this
makes it possible for vendoring tool to vendor tuf. Fix all references
to <module> in the code.

Also fix import orders so tuf internal imports are last.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Make 'log' imports vendoring-compatible

Use "from tuf import <module>" instead of "import tuf.<module>": this
makes it possible for vendoring tool to vendor tuf. Fix all references
to <module> in the code.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
updater: rename mirrors variables

Make sure mirrors is not used as variable name (so it can be used for
the module import name later).

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Make 'mirrors' imports vendoring compatible

Use "from tuf import <module>" instead of "import tuf.<module>": this
makes it possible for vendoring tool to vendor tuf. Fix all references
to <module> in the code.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Make 'roledb' imports vendoring-compatible

Use "from tuf import <module>" instead of "import tuf.<module>": this
makes it possible for vendoring tool to vendor tuf. Fix all references
to <module> in the code.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Make 'settings' imports vendoring-compatible

Use "from tuf import <module>" instead of "import tuf.<module>": this
makes it possible for vendoring tool to vendor tuf. Fix all references
to <module> in the code.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Make 'sig' imports vendoring-compatible

Use "from tuf import <module>" instead of "import tuf.<module>": this
makes it possible for vendoring tool to vendor tuf. Fix all references
to <module> in the code.

Remove one unused import.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Fix securesystemslib.exceptions imports

Make the import compatible with vendoring tool and alias the import so
it does not clash with the local module. Fix all references
to the module in the code.

Remove a related repo.py comment that was badly duplicated from module
docstring.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Make 'updater' import vendoring-compatible

Make the updater imports compatible with vendoring tool by importing
the Updater class directly (don't import the whole module to avoid the
clash with the obvious variable name 'updater').

Also update the example: This is not required in the clients but tuf
source code will be vendored and this import line (even though in a
comment) might trigger an error in future vendoring tool releases.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Fix securesystemslib.formats imports

Make the import compatible with vendoring tool and alias the import so
it does not clash with the local module. Fix all references
to the module in the code.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Fix securesystemslib.keys imports

Make them compatible with vendoring, use
  from securesystemslib import keys as sslib_keys
to have the same style as other securesystemslib imports.

Note that developer_tool already used a
   from securesystemslib.keys import ...
for some functions so that style was used consistently there.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Fix securesystemslib.interface imports

Make them compatible with vendoring, use
  from securesystemslib import interface as sslib_interface
to have the same style as other securesystemslib imports.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Fix securesystemslib.util imports

Make the import compatible with vendoring tool and alias the import so
it does not clash with the local module. Fix all references
to the module in the code.

In one instance import a specific function to avoid a more complex
redirection in the code.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Fix securesystemslib.storage imports

Make them compatible with vendoring, use
  from securesystemslib import storage as sslib_storage
to have the same style as other securesystemslib imports.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Fix securesystemslib.hash imports

Make them compatible with vendoring, use
  from securesystemslib import hash as sslib_hash
to have the same style as other securesystemslib imports (and to avoid
potential conflict with system hash()).

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Fix securesystemslib.settings imports

Make the import compatible with vendoring tool and alias the import so
it does not clash with the local module. Fix all references
to the module in the code.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Move six imports to 3rd party section

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
imports: Remove unused imports

The linter now understands our imports (yay), and complains a lot (boo):
* Remove really unused imports
* disable lints for tuf.log and securesystemslib imports: these imports
  have logging side-effects (they set default loggers for tuf and
  securesystemslib respectively) and I'm cautious about just removing
  them

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

joshuagl commented on tuf/keydb.py:
```diff
@@ -344,7 +344,7 @@ def get_key(keyid, repository_name='default'):
 
   except KeyError as error:
-    six.raise_from(tuf.exceptions.UnknownKeyError('Key: ' + keyid), error)
+    six.raise_from(exceptions.UnknownKeyError('Key: ' + keyid), error)
```
Looks like this was missed in the prior commit? 😃

joshuagl commented on tuf/developer_tool.py:
```diff
@@ -251,12 +251,12 @@ def write(self, write_partial=False):
     securesystemslib.formats.BOOLEAN_SCHEMA.check_match(write_partial)
 
-    # At this point the tuf.keydb and tuf.roledb stores must be fully
+    # At this point the tuf.keydb and roledb stores must be fully
```
should a prior commit have changed tuf.keydb -> keydb ?

jku replied:
it looks like I have two commits with keydb changes, the latter one seems to handle these. Will combine them.

joshuagl commented on tuf/developer_tool.py:
```diff
@@ -310,7 +310,7 @@ def add_verification_key(self, key, expires=None):
 
       <Side Effects>
-        The role's entries in 'tuf.keydb.py' and 'tuf.roledb.py' are updated.
+        The role's entries in 'tuf.keydb.py' and 'roledb' are updated.
```
should a prior commit have changed tuf.keydb.py -> keydb ?

joshuagl commented on tuf/repository_lib.py:
```diff
@@ -1195,7 +1195,7 @@ def generate_root_metadata(version, expiration_date, consistent_snapshot,
   """
   <Purpose>
-    Create the root metadata.  'tuf.roledb.py' and 'tuf.keydb.py'
+    Create the root metadata.  'roledb' and 'tuf.keydb.py'
```
tuf.keydb.py -> keydb, in a prior commit?

joshuagl commented on tuf/repository_lib.py:
```diff
@@ -1224,10 +1224,10 @@ def generate_root_metadata(version, expiration_date, consistent_snapshot,
 
   <Side Effects>
-    The contents of 'tuf.keydb.py' and 'tuf.roledb.py' are read.
+    The contents of 'tuf.keydb.py' and 'roledb' are read.
```
tuf.keydb.py -> keydb, in a prior commit?

joshuagl commented on tuf/repository_tool.py:
```diff
@@ -314,7 +314,7 @@ def writeall(self, consistent_snapshot=False, use_existing_fileinfo=False):
     securesystemslib.formats.BOOLEAN_SCHEMA.check_match(consistent_snapshot)
 
-    # At this point, tuf.keydb and tuf.roledb must be fully populated,
+    # At this point, tuf.keydb and roledb must be fully populated,
```
tuf.keydb -> keydb, in a prior commit?

joshuagl commented on tuf/repository_tool.py:
```diff
@@ -721,8 +721,7 @@ def add_verification_key(self, key, expires=None):
     <Side Effects>
-      The role's entries in 'tuf.keydb.py' and 'tuf.roledb.py' are
-      updated.
+      The role's entries in 'tuf.keydb.py' and 'roledb' are updated.
```
tuf.keydb.py -> keydb, in a prior commit?

joshuagl commented on tuf/repository_tool.py:
```diff
@@ -883,7 +882,7 @@ def load_signing_key(self, key):
 
     <Side Effects>
-      Updates the role's 'tuf.keydb.py' and 'tuf.roledb.py' entries.
+      Updates the role's 'tuf.keydb.py' and 'roledb' entries.
```
tuf.keydb.py -> keydb, in a prior commit?

joshuagl commented on tuf/repository_tool.py:
```diff
@@ -2353,7 +2352,7 @@ def delegate(self, rolename, public_keys, paths, threshold=1,
     <Side Effects>
       A new Target object is created for 'rolename' that is accessible to the
       caller (i.e., targets.<rolename>).  The 'tuf.keydb.py' and
```
tuf.keydb.py -> keydb, in a prior commit?

joshuagl commented on tuf/sig.py:
```diff
@@ -71,14 +71,14 @@ def get_signature_status(signable, role=None, repository_name='default',
 
     * bad -- Invalid signature
     * good -- Valid signature from key that is available in 'tuf.keydb', and is
```
tuf.keydb -> keydb, in a prior commit?

joshuagl commented on tuf/sig.py:
```diff
@@ -71,14 +71,14 @@ def get_signature_status(signable, role=None, repository_name='default',
-      authorized for the passed role as per 'tuf.roledb' (authorization may be
+      authorized for the passed role as per 'roledb' (authorization may be
       overwritten by passed 'keyids').
     * unknown -- Signature from key that is not available in 'tuf.keydb', or if
```
tuf.keydb -> keydb, in a prior commit?

joshuagl commented on tuf/sig.py:
```diff
@@ -71,14 +71,14 @@ def get_signature_status(signable, role=None, repository_name='default',
       'role' is None.
     * unknown signing schemes -- Signature from key with unknown signing
       scheme.
     * untrusted -- Valid signature from key that is available in 'tuf.keydb',
```
tuf.keydb -> keydb, in a prior commit?

joshuagl commented on tuf/download.py:
```diff
@@ -357,14 +358,14 @@ def _download_fixed_amount_of_data(response, temp_file, required_length):
+      if average_download_speed < settings.MIN_AVERAGE_DOWNLOAD_SPEED:
         logger.debug('The average download speed dropped below the minimum'
-          ' average download speed set in tuf.settings.py.')
+          ' average download speed set in tuf.settings.')
```
We're somewhat inconsistent across this PR on whether we refer to modules – in comments and log messages – as tuf.module, or just module. Let's choose one and stick to it.

joshuagl commented on tuf/client/updater.py:
```diff
@@ -129,6 +129,8 @@
 import copy
 import warnings
 
+from securesystemslib import exceptions as sslib_exceptions
```
Given that the exceptions module is one of the few places in sslib where we have classes, wouldn't it be cleaner to import only the exceptions we want/need? Or do some of the conflicting exception names between sslib and tuf prevent this?

jku replied:
This would probably be a good idea

jku replied:
I've not done this for now... there are cases where clashes are definitely possible and a couple where they actually exist right now: the main issue is the generic "Error" exception that both projects use.

It's an idea worth exploring but I think not as part of this PR.

joshuagl replied:
Sounds reasonable, thanks for looking into it.

joshuagl commented on tuf/log.py:
```diff
@@ -269,7 +271,7 @@ def set_console_log_level(log_level=_DEFAULT_CONSOLE_LOG_LEVEL):
 
   <Exceptions>
-    securesystems.exceptions.Error, if the 'log.py' console handler has not
+    securesystemslib.exceptions.Error, if the 'log.py' console handler has not
```
log.py -> log?

jku replied:
the whole file is full of this naming, decided not to touch it

joshuagl commented on tuf/scripts/repo.py:
```diff
@@ -147,17 +147,19 @@
 import fnmatch
 
+# 'pip install securesystemslib[crypto,pynacl]' is required for the CLI,
+# which installs the cryptography and pynacl.
```
"which installs the cryptography and pynacl dependedencies"?

joshuagl commented on tuf/client/updater.py:
```diff
@@ -130,6 +130,7 @@
 import warnings
 
 from securesystemslib import exceptions as sslib_exceptions
+from securesystemslib import formats as sslib_formats
```
Can we just import the formats we care about?
`from securesystemslib import PATH_SCHEMA, NAME_SCHEMA, ...` ?

jku replied:
probably reasonable, will test this for sslib_formats and sslib_exceptions

jku replied:
shoot, I had missed this suggestion in this pass. This is not done yet...

jku replied:
Decided not to do this after all: in some cases this approach would work but there are a lot of uses in repository_tool, repository_lib, keydb and roledb . Some of them would end up with clashes (at least exceptions.Error)... The schemas probably would work but I'm not sure the amount of work is reasonable if we want to change all uses.

I think I would prefer leaving this outside of the PR.

joshuagl replied:
Sounds reasonable to me.

joshuagl commented on tuf/developer_tool.py:
```diff
@@ -75,7 +74,8 @@
 from securesystemslib.keys import (
-    format_keyval_to_metadata)
+    format_keyval_to_metadata,
+    format_metadata_to_key)
```
Curious why you chose this pattern here, importing specific functions, but not in updater?

jku replied:
oh just because developer_tool was already using the "from x import y" pattern for keys (just not consistently). It's likely that other places would benefit from doing that but in those cases I would have had more new local name conflicts to worry about (when I import the module, I can just check the module name is not used)

joshuagl reviewed (commented):
Thank you for tackling this @jku! 

FYI the series has multiple "imports: Fix securesystemslib.exceptions imports" patches. They aren't all modifying exceptions. Similarly with "imports: Fix securesystemslib.interface imports".

> This is a draft since I'll have to do something similar in securesystemslib before I can test that it all actually works with vendoring... I am also going to go through this PR again to make sure there are no typos and to improve commit messages.

Should have paid more attention to this before reviewing commit by commit, I hope the numerous comments I've left are useful.
 

> * as result securesystemslib inports are now like `from securesystemslib import formats as sslib_formats` in most cases. I don't particularly like it and am willing to consider alternatives

I'm not opposed to this, though I should point out that we can avoid many of the sslib_* imports by importing only the names we care about, you've done this in a couple of places – importing only the functions from a securesystemslib module that the tuf module is using – I assume expediency is the main reason you didn't do this throughout?

> * The patch is surprisingly large, ~900 lines. Roughly 350 are needed for the client vendoring but I modified the whole code base to at least be consistent. The tests have not been touched.

Completely agree with this rationale.

lukpueh commented:
Thanks a ton for taking a stab at this thankless task, and especially for the thorough separation of commits! As requested, I only skimmed the patches here and in secure-systems-lab/securesystemslib#316. Let me know when you want a full review.

Here are some thoughts regarding your notes from above.

> * tuf modules are now imported with this style `from tuf import download`. 

I'd like to add that this does not only fix the vendoring issues but also aligns with our [style recommendation](https://github.com/secure-systems-lab/code-style-guidelines/blame/28bccecd4ff0783e9e0395c8e6d7cbc9e532bc34/python.md#L55-L56), even though the [Google guide accepts both](https://github.com/secure-systems-lab/code-style-guidelines/pull/21#discussion_r533310544).


> * as result securesystemslib inports are now like `from securesystemslib import formats as sslib_formats` in most cases. I don't particularly like it and am willing to consider alternatives

I'm fine with this approach. IMO the secondary goal (besides making vendoring work) should be to help the reader attribute function or other names to modules or packages. And `sslib_interface.import_publickeys_from_file` seems to serve that purpose better than `interface.import_publickeys_from_file`.

> * The import lists that we're changed have been sorted so tuf imports are last

Excellent, this too is a style guide recommendation.

* The patch is surprisingly large, ~900 lines. Roughly 350 are needed for the client vendoring but I modified the whole code base to at least be consistent. The tests have not been touched.

👍 


Do you think it's worth to add an ADR for the rationale in this PR? Or to amend our style guide? Specifically in regards to:

`from package import module` vs. `from package import module as xyz_module` vs. `from package.module import name`

jku commented:
> To you think it's worth to add an ADR for the rationale in this PR? Or to amend our style guide? Specifically in regards to:
> 
> `from package import module` vs. `from package import module as xyz_module` vs. `from package.module import name`

Eh, I think it maybe depends... So far I've mostly used the first (and second where confusion was more likely) but the third one makes sense for some modules. I've avoided the third option for now mostly to keep the changes to a minimum and so I only had to worry about one new local name (the module name) and not multiple as might be the case if we import individual things from modules.

Related thought (this might not be a good idea, just something I'm wondering), for some of our most common imports I've been thinking: If TUF uses e.g. schemas from `tuf.formats` and `securesystemslibs.formats` wouldn't it make sense for `tuf.formats` to expose the relevant items from `securesystemslibs.formats`? This way there would be one place (tuf.formats) that takes care of possible conflicts and other TUF code could just `import tuf.formats` or individual items from there (module name not really needed because now there's no confusion over where they come from). Same goes to a degree for exceptions -- where it's a bit weird that code that does not call securesystemslib code might have to import `securesystemslib.exceptions` to handle errors.

lukpueh commented:
> Related thought (this might not be a good idea, just something I'm wondering), for some of our most common imports I've been thinking: If TUF uses e.g. schemas from tuf.formats and securesystemslibs.formats wouldn't it make sense for tuf.formats to expose the relevant items from securesystemslibs.formats? 

I generally think that's a good idea, but given that we want to get away from schemas (https://github.com/secure-systems-lab/securesystemslib/issues/183) maybe not worth the effort?

> Same goes to a degree for exceptions -- where it's a bit weird that code that does not call securesystemslib code might have to import securesystemslib.exceptions to handle errors.

I agree that it's weird. But I think the solution should be a revision of the exception taxonomy (https://github.com/theupdateframework/tuf/issues/1131, secure-systems-lab/securesystemslib#191). AFAICS there is no big win in defining any tuf exceptions in securesystemslib.

Commit:
imports: Fix urllib3 exception import

Make it compatible with vendoring: import the exception only to avoid
having to rename the module locally.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
requests_fetcher: Move 'tuf' import from download

requests_fetcher uses tuf.__version__ for user-agent, move the import to
the correct file.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
I've marked this ready for review. 

I think this is the one question that hasn't been answered (in this issue):
> I'm not opposed to this, though I should point out that we can avoid many of the sslib_* imports by importing only the names we care about, you've done this in a couple of places – importing only the functions from a securesystemslib module that the tuf module is using – I assume expediency is the main reason you didn't do this throughout?

I think the places where I did it were already using this format for some other formats from this module (so I was aiming for least disruption to existing code). Otherwise I tried to follow the guideline of importing modules, not module contents. It's true that in some instances the latter would work better: schemas are an example but modifying them is a lot of work that might not be 100% trivial and that I'd rather leave for a follow up if needed

lukpueh commented on tuf/requests_fetcher.py:
```diff
@@ -14,7 +14,7 @@
 
 from tuf import exceptions
-import tuf.settings
+from tuf import settings
```
Nit pick: I think the patch in this file should have gone into 7841008a. (Note that I mostly point this out so that you see that I actually read the diff 😄 . Feel free to ignore.)

lukpueh reviewed (approved):
Thanks again for doing this dull deed, @jku! Here is how I reviewed your PR:

I have **not** verified if this patch
- indeed makes tuf vendoring-compatible,
- is complete,
- doesn't shadow any names.

Also, I have **not** commented on inconsistent adoption of new module/sybmol paths in docstrings, most notably in `<Exceptions>`, or `Raises:` sections. I noticed some and briefly considered pointing them out, but I'm unsure if it is worth the effort.


What **I did do** is I
- went through each commit,
- read each diff line by line, and
- checked that each diff corresponds to the commit message.
- I also made sure you addressed @joshuagl's comments.

Based on that, your PR LGTM.

If that's enough, I'd say we merge, after you have rebased and fixed the conflict in `tuf/api/metadata.py`. 

ps: I'd also like to land this before https://github.com/theupdateframework/tuf/pull/1314 because that way it will be easier to resolve conflicts (nothing big, I checked, but still).

Commit:
Make requests_fetcher import vendoring compatible

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
metadata: Make isort happy and bundle imports

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
Changes since last push:
* rebased to develop
* fixed rebase issues in "imports: Make 'formats' imports vendoring-compatible" 
* fixed rebase issues in "imports: Make 'exceptions' imports vendoring-compatible" 
* Added a commit for a missed requests_fetcher import
* Added a commit to bundle imports like isort likes them

lukpueh commented:
Still LGTM! :)
